File: solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/Middleware/Telemetry/TelemetryExtensions.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;

namespace Microsoft.Bot.Solutions.Middleware.Telemetry
{
    /// <summary>
    /// This class hosts the extension methods for telemetry purposes to add the basic logging properties before taking the appropriate action.
    /// </summary>
    public static class TelemetryExtensions
    {
        private const string NoActivityId = "no activity id";
        private const string NoChannelId = "no channel id";
        private const string NoConversationId = "no conversation id";
        private const string NoDialogId = "no dialog id";
        private const string NoUserId = "no user id";

        public static void TrackEventEx(this IBotTelemetryClient telemetryClient, string eventName, Activity activity, string dialogId = null, IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null)
        {
            telemetryClient.TrackEvent(eventName, GetFinalProperties(activity, dialogId, properties), metrics);
        }

        public static void TrackExceptionEx(this IBotTelemetryClient telemetryClient, Exception exception, Activity activity, string dialogId = null, IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null)
        {
            telemetryClient.TrackException(exception, GetFinalProperties(activity, dialogId, properties), metrics);
        }

        public static void TrackTraceEx(this IBotTelemetryClient telemetryClient, string message, Severity severityLevel, Activity activity, IDictionary<string, string> properties, string dialogId = null)
        {
            telemetryClient.TrackTrace(message, severityLevel, GetFinalProperties(activity, dialogId, properties));
        }

        private static IDictionary<string, string> GetFinalProperties(Activity activity, string dialogId, IDictionary<string, string> properties = null)
        {
            var finalProperties = new Dictionary<string, string>();

            finalProperties.Add(TelemetryConstants.ActiveDialogIdProperty, dialogId ?? NoDialogId);

            finalProperties.Add(TelemetryConstants.ActivityIDProperty, activity?.Id ?? NoActivityId);
            finalProperties.Add(TelemetryConstants.ChannelIdProperty, activity?.ChannelId ?? NoChannelId);
            finalProperties.Add(TelemetryConstants.ConversationIdProperty, activity?.Conversation?.Id ?? NoConversationId);
            finalProperties.Add(TelemetryConstants.UserIdProperty, activity?.From?.Id ?? NoUserId);

            if (properties != null && properties.Count > 0)
            {
                foreach (var property in properties)
                {
                    var key = property.Key;
                    var value = property.Value;

                    if (finalProperties.ContainsKey(key))
                    {
                        finalProperties[key] = value;
                    }
                    else
                    {
                        finalProperties.Add(key, value);
                    }
                }
            }

            return finalProperties;
        }
    }
}

File: solutions/Virtual-Assistant/src/typescript/bot-solution/src/middleware/telemetry/telemetryExtensions.ts
import { Activity, BotTelemetryClient, Severity } from 'botbuilder';
import { TelemetryConstants } from './telemetryLoggerMiddleware';

export namespace TelemetryExtensions {
    const noActivityId: string = 'no activity id';
    const noChannelId: string = 'no channel id';
    const noConversationId: string = 'no conversation id';
    const noDialogId: string = 'no dialog id';
    const noUserId: string = 'no user id';

    export function trackEventEx(client: BotTelemetryClient, eventName: string, activity: Partial<Activity>, dialogId?: string,
                                 properties?: { [key: string]: string }, metrics?: { [key: string]: number }): void {
        client.trackEvent({
            name: eventName,
            properties: getFinalProperties(activity, dialogId, properties),
            metrics: metrics
        });
    }

    export function trackExceptionEx(client: BotTelemetryClient, exception: Error, activity: Partial<Activity>, dialogId?: string,
                                     properties?: { [key: string]: string }, metrics?: { [key: string]: number }): void {
        client.trackException({
            exception: exception,
            properties: getFinalProperties(activity, dialogId, properties),
            measurements: metrics
        });
    }

    export function trackTraceEx(client: BotTelemetryClient, message: string, severity: Severity, activity: Partial<Activity>,
                                 dialogId?: string, properties?: { [key: string]: string }): void {
        client.trackTrace({
            message: message,
            properties: getFinalProperties(activity, dialogId, properties),
            severityLevel: severity
        });
    }

    function getFinalProperties(activity: Partial<Activity>, dialogId?: string,
                                properties?: { [key: string]: string }): { [key: string]: string } {
        const finalProps: { [key: string]: string } = {};

        finalProps[TelemetryConstants.activeDialogIdProperty] = dialogId || noDialogId;

        finalProps[TelemetryConstants.activityIDProperty] = activity.id || noActivityId;
        finalProps[TelemetryConstants.channelIdProperty] = activity.channelId || noChannelId;
        finalProps[TelemetryConstants.conversationIdProperty] = activity.conversation ? activity.conversation.id : noConversationId;
        finalProps[TelemetryConstants.userIdProperty] = activity.from ? activity.from.id : noUserId;

        if (properties) {
            Object.assign(finalProps, properties);
        }

        return finalProps;
    }
}
